{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "285\n"
     ]
    }
   ],
   "source": [
    "#find specific species\n",
    "def read_and_process_file(input_file_path, output_file_path):\n",
    "    with open(input_file_path, 'r') as file, open(output_file_path, 'w') as output_file:\n",
    "        next(file)  # Skip the first line\n",
    "        n=0\n",
    "        for line in file:\n",
    "            parts = line.strip().split(', ')\n",
    "            if len(parts) >= 3:\n",
    "                name, purity, bacteria = parts[0], float(parts[1]), parts[2]\n",
    "                if bacteria == \"species\": #the species you want to analysis\n",
    "                    n=n+1\n",
    "                    original_file_name = name\n",
    "                    output_file.write(original_file_name  + '\\n')\n",
    "        print(n)\n",
    "input_file_path = 'purity_minimap.txt'  # Replace with your actual input file path\n",
    "output_file_path = 'fastq_list_species.txt'  # Replace with your desired output file path\n",
    "\n",
    "read_and_process_file(input_file_path, output_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#purity plot for the specific species\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def extract_purity_and_plot_histogram(file_path):\n",
    "    purities = []\n",
    "\n",
    "    with open(file_path, 'r') as file:\n",
    "        next(file)  # Skip the first line\n",
    "        for line in file:\n",
    "            parts = line.strip().split(', ')\n",
    "            if len(parts) >= 3 and parts[2] == \"species\":\n",
    "                purity = float(parts[1])\n",
    "                purities.append(purity)\n",
    "\n",
    "    # Plotting the histogram\n",
    "    bins = np.arange(0, 105, 5)\n",
    "    plt.hist(purities, bins=bins, edgecolor='black')\n",
    "    plt.title('Purity Distribution for species')\n",
    "    plt.xlabel('Purity (%)')\n",
    "    plt.ylabel('Frequency')\n",
    "    plt.show()\n",
    "\n",
    "file_path = 'purity_minimap2.txt'  # Replace with your actual file path\n",
    "extract_purity_and_plot_histogram(file_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should turn to coverage.sh before running the following cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def read_folder_names(file_path):\n",
    "    folder_names = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            folder_name = line.strip().replace('_minimap2_output.txt', '')\n",
    "            folder_names.append(folder_name)\n",
    "    return folder_names\n",
    "\n",
    "\n",
    "#calculate_theoretical_coverage\n",
    "def calculate_theoretical_coverage(fastq_path, reference_genome_size):\n",
    "    base_count = 0\n",
    "    line_counter = 0  # track record\n",
    "    \n",
    "    with open(fastq_path, 'r') as file:\n",
    "        for line in file:\n",
    "            stripped_line = line.strip()\n",
    "            \n",
    "            if line.startswith('@'):\n",
    "                line_counter = 1  # first line\n",
    "            else:\n",
    "                line_counter += 1\n",
    "                \n",
    "            # only caculate the second line\n",
    "            if line_counter == 2 and stripped_line:\n",
    "                base_count += len(stripped_line)\n",
    "    \n",
    "    return base_count / reference_genome_size\n",
    "\n",
    "def get_theoretical_coverage(folder_path, folder_names, reference_genome_size):\n",
    "    theoretical_coverages = {}\n",
    "    for folder_name in folder_names:\n",
    "        for file in os.listdir(folder_path):\n",
    "            if file.endswith('.fastq') and folder_name in file:\n",
    "                coverage = calculate_theoretical_coverage(os.path.join(folder_path, file), reference_genome_size)\n",
    "                theoretical_coverages[folder_name] = coverage\n",
    "                break\n",
    "    return theoretical_coverages\n",
    "\n",
    "# actual_coverage\n",
    "def get_actual_coverage(new_path, folder_names):\n",
    "    actual_coverages = {}\n",
    "    for folder_name in folder_names:\n",
    "        coverage_file = os.path.join(new_path, folder_name, 'coverage_percentage.txt')\n",
    "        if os.path.exists(coverage_file):\n",
    "            with open(coverage_file, 'r') as file:\n",
    "                coverage = float(file.read().strip())\n",
    "                actual_coverages[folder_name] = coverage\n",
    "    return actual_coverages\n",
    "\n",
    "\n",
    "def plot_coverages(theoretical_coverages, actual_coverages):\n",
    "    x = [theoretical_coverages[name] for name in theoretical_coverages]\n",
    "    y = [actual_coverages[name] for name in actual_coverages if name in theoretical_coverages]\n",
    "\n",
    "    plt.scatter(x, y)\n",
    "    plt.xlabel('Sequencing Depth')\n",
    "    plt.ylabel('Coverage(%)')\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def plot_combined_coverages(theoretical_coverages_1, actual_coverages_1, theoretical_coverages_2, actual_coverages_2,theoretical_coverages_3, actual_coverages_3,theoretical_coverages_4, actual_coverages_4):\n",
    "    # first group, blue\n",
    "    x1 = [theoretical_coverages_1[name] for name in theoretical_coverages_1]\n",
    "    y1 = [actual_coverages_1[name] for name in actual_coverages_1 if name in theoretical_coverages_1]\n",
    "    plt.scatter(x1, y1, color='blue', marker='o', label='Bacillus', s = 10)\n",
    "\n",
    "    # second group, red\n",
    "    x2 = [theoretical_coverages_2[name] for name in theoretical_coverages_2]\n",
    "    y2 = [actual_coverages_2[name] for name in actual_coverages_2 if name in theoretical_coverages_2]\n",
    "    plt.scatter(x2, y2, color='red', marker='x', label='Ecoli', s = 10)\n",
    "\n",
    "    # third group, green\n",
    "    x3 = [theoretical_coverages_3[name] for name in theoretical_coverages_3]\n",
    "    y3 = [actual_coverages_3[name] for name in actual_coverages_3 if name in theoretical_coverages_3]\n",
    "    plt.scatter(x3, y3, color='green', marker='^', label='Staphylococcus', s = 10)\n",
    "\n",
    "    # fourth group, black\n",
    "    x4 = [theoretical_coverages_4[name] for name in theoretical_coverages_4]\n",
    "    y4 = [actual_coverages_4[name] for name in actual_coverages_4 if name in theoretical_coverages_4]\n",
    "    plt.scatter(x4, y4, color='black', marker='s', label='Klebsiella', s = 10)\n",
    "\n",
    "    plt.xlabel('Sequencing Depth')\n",
    "    plt.ylabel('Coverage(%)')\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "\n",
    "def write_results_to_file(depths,coverages, output_path):\n",
    "    with open(output_path, \"w\") as file:\n",
    "        file.write(\"Filename, coverage (%), depth\\n\")\n",
    "        for file_name, coverage in coverages.items():\n",
    "            # Convert coverage and depth to float if they are not already\n",
    "            depth = depths[file_name]\n",
    "            try:\n",
    "                coverage_float = float(coverage)\n",
    "                depth_float = float(depth)\n",
    "            except ValueError:\n",
    "                raise ValueError(f\"Coverage or depth could not be converted to float: coverage={coverage}, depth={depth}\")\n",
    "            \n",
    "            file.write(f\"{file_name}, {coverage_float:.8f}, {depth_float:.8f}\\n\")\n",
    "\n",
    "\n",
    "def main():\n",
    "    # total bases in reference genome\n",
    "    reference_genome_size_Bacillus = 4224687  # change to real reference genome used\n",
    "    reference_genome_size_Ecoli = 4567887  # change to real reference genome used\n",
    "    reference_genome_size_Staphylococcus = 2821361  # change to real reference genome used\n",
    "    reference_genome_size_Klebsiella = 5333942  # change to real reference genome used\n",
    "\n",
    "    #input file, you should change to real file\n",
    "    file_path_1 = 'fastq_list_Bacillus.txt'  \n",
    "    folder_names_1 = read_folder_names(file_path_1)\n",
    "\n",
    "    file_path_2 = 'fastq_list_Ecoli.txt'  \n",
    "    folder_names_2 = read_folder_names(file_path_2)\n",
    "\n",
    "    file_path_3 = 'fastq_list_Staphylococcus.txt' \n",
    "    folder_names_3 = read_folder_names(file_path_3)\n",
    "\n",
    "    file_path_4 = 'fastq_list_Klebsiella.txt' \n",
    "    folder_names_4 = read_folder_names(file_path_4)\n",
    "\n",
    "    # calculate_theoretical_coverage\n",
    "    folder_path = './bbmap_output_many_seqs' \n",
    "    theoretical_coverages_1 = get_theoretical_coverage(folder_path, folder_names_1, reference_genome_size_Bacillus)\n",
    "    theoretical_coverages_2 = get_theoretical_coverage(folder_path, folder_names_2, reference_genome_size_Ecoli)\n",
    "    theoretical_coverages_3 = get_theoretical_coverage(folder_path, folder_names_3, reference_genome_size_Staphylococcus)\n",
    "    theoretical_coverages_4 = get_theoretical_coverage(folder_path, folder_names_4, reference_genome_size_Klebsiella)\n",
    "\n",
    "    # get actual coverage, you should change to real file\n",
    "    new_path_1 = './coverage/Bacillus/'  \n",
    "    actual_coverages_1 = get_actual_coverage(new_path_1, folder_names_1)\n",
    "\n",
    "    new_path_2 = './coverage/Ecoli/' \n",
    "    actual_coverages_2 = get_actual_coverage(new_path_2, folder_names_2)\n",
    "\n",
    "    new_path_3 = './coverage/Staphylococcus/' \n",
    "    actual_coverages_3 = get_actual_coverage(new_path_3, folder_names_3)\n",
    "    \n",
    "    new_path_4 = './coverage/Klebsiella/'  \n",
    "    actual_coverages_4 = get_actual_coverage(new_path_4, folder_names_4)\n",
    " \n",
    "    # scatter plot\n",
    "    plot_coverages(theoretical_coverages_1, actual_coverages_1)\n",
    "    plot_coverages(theoretical_coverages_2, actual_coverages_2)\n",
    "    plot_coverages(theoretical_coverages_3, actual_coverages_3)\n",
    "    plot_coverages(theoretical_coverages_4, actual_coverages_4)\n",
    "    plot_combined_coverages(theoretical_coverages_1, actual_coverages_1, theoretical_coverages_2, actual_coverages_2,theoretical_coverages_3, actual_coverages_3,theoretical_coverages_4, actual_coverages_4)\n",
    "\n",
    "    output_path = \"./coverage/coverage_Bacillus.txt\"\n",
    "    \n",
    "    write_results_to_file(theoretical_coverages_1,actual_coverages_1, output_path)\n",
    "    output_path = \"./coverage/coverage_Ecoli.txt\"\n",
    "    write_results_to_file(theoretical_coverages_2,actual_coverages_2, output_path)\n",
    "\n",
    "    output_path = \"./coverage/coverage_Staphylococcus.txt\"\n",
    "    write_results_to_file(theoretical_coverages_3,actual_coverages_3, output_path)\n",
    "\n",
    "    output_path = \"./coverage/coverage_Klebsiella.txt\"\n",
    "    write_results_to_file(theoretical_coverages_4,actual_coverages_4, output_path)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
